lib/matplotlib/axes/_axes.py

```diff
@@ -4389,6 +4413,19 @@ def do_patch(xs, ys, **kwargs):
                 flier_y = stats['fliers']
                 fliers.append(do_plot(flier_x, flier_y, **flier_kw))
 
+        # Set legend labels
+        if label:
+            box_or_med = boxes if showbox and patch_artist else medians
+            if cbook.is_scalar_or_string(label):
+                # assign the label only to the first box
+                box_or_med[0].set_label(label)
+            else:  # label is a sequence
+                if len(box_or_med) != len(label):
+                    raise ValueError("There must be an equal number of legend"
+                                     " labels and boxplots.")
```

Comment:
This check should probably be in the "input validation" section, and reuse `datashape_message`.